src/sys/windows.rs

```diff
@@ -735,6 +736,52 @@ pub(crate) fn from_in6_addr(addr: in6_addr) -> Ipv6Addr {
     Ipv6Addr::from(*unsafe { addr.u.Byte() })
 }
 
+/// This type is not yet in winapi.
+#[repr(C)]
+#[allow(non_camel_case_types)]
+struct sockaddr_un {
+    pub sun_family: sa_family_t,
+    pub sun_path: [u8; 108],
+}
+
+#[allow(unused_unsafe)] // TODO: replace with `unsafe_op_in_unsafe_fn` once stable.
+pub(crate) fn unix_sockaddr(path: &Path) -> io::Result<SockAddr> {
+    unsafe {
+        SockAddr::init(|storage, len| {
+            // Safety: `SockAddr::init` zeros the address, which is a valid
+            // representation.
+            let storage: &mut crate::sys::sockaddr_un = unsafe { &mut *storage.cast() };
+            let len: &mut socklen_t = unsafe { &mut *len };
+
+            let bytes = path
+                .to_str()
```

Comment:
What we really need here is a `Path`/`OsStr` -> `&[u8]` method don't we? For Unix this already exists. For Windows all the way down this is a `Wtf8` which contains one field: `bytes: [u8]`, so it shouldn't be too hard to add.

Reply:
Internally Windows uses a 16-bit representation and converts sun_path to it. Unfortunately, this conversion will reject strings with unmatched Unicode surrogates, meaning that it will only accept UTF-8 strings, not the WTF-8 superset, even though the 16-bit representation for Windows paths _does_ allow unmatched surrogates.

So even if this method existed and we used it here, it would just defer the failure to the point of the bind/connect call.

Reply:
> So even if this method existed and we used it here, it would just defer the failure to the point of the bind/connect call.

That's fine we don't check if the path is valid for Unix either (except for correct usage, i.e. null terminated), we just pass it along to the OS.

Reply:
I've clarified this in a comment.

Reply:
We should also open an issue with the Rust repo to allow use access to the underlying bytes.

Reply:
I found an existing draft PR that discusses this, I'll link it. https://github.com/rust-lang/rust/pull/95290